pkg/webhook/cspc.go

```diff
@@ -0,0 +1,178 @@
+		}
+		return response
+	}
+
+	if ok, msg := cspcValidation(&cspc); !ok {
+		response.Allowed = false
+		response.Result = &metav1.Status{
+			Message: fmt.Sprintf("invalid cspc specification: %s", msg),
+		}
+		return response
+	}
+	return response
+}
+
+func cspcValidation(cspc *apis.CStorPoolCluster) (bool, string) {
+	if len(cspc.Spec.Pools) == 0 {
+		return false, fmt.Sprintf("pools in cspc should have at least one item")
+	}
+
+	for _, pool := range cspc.Spec.Pools {
+		ok, msg := poolSpecValidation(&pool)
+		if !ok {
+			return false, fmt.Sprintf("invalid cspc spec: %s", msg)
+		}
+	}
+	return true, ""
+}
+
+func poolSpecValidation(pool *apis.PoolSpec) (bool, string) {
+	if pool.NodeSelector == nil || len(pool.NodeSelector) == 0 {
+		return false, "nodeselector should not be empty"
+	}
+	if len(pool.RaidGroups) == 0 {
+		return false, "at least one raid group should be present on pool spec"
+	}
+	// TODO : Add validation for pool config
+	// Pool config will require mutating webhooks also.
+	for _, raidGroup := range pool.RaidGroups {
+		ok, msg := raidGroupValidation(&raidGroup, &pool.PoolConfig)
+		if !ok {
+			return false, msg
+		}
+	}
+
+	return true, ""
+}
+
+func raidGroupValidation(raidGroup *apis.RaidGroup, pool *apis.PoolConfig) (bool, string) {
+	if raidGroup.Type == "" && pool.DefaultRaidGroupType == "" {
+		return false, fmt.Sprintf("any one type at raid group or default raid group type be specified ")
+	}
+	if _, ok := apis.SupportedPRaidType[apis.PoolType(raidGroup.Type)]; !ok {
+		return false, fmt.Sprintf("unsupported raid type '%s' specified", apis.PoolType(raidGroup.Type))
+	}
+
+	if len(raidGroup.BlockDevices) == 0 {
+		return false, fmt.Sprintf("number of block devices honouring raid type should be specified")
+	}
+
+	if raidGroup.Type != "stripe" {
+		if len(raidGroup.BlockDevices) != apis.SupportedPRaidType[apis.PoolType(raidGroup.Type)] {
+			return false, fmt.Sprintf("number of block devices honouring raid type should be specified")
+		}
+	} else {
+		if len(raidGroup.BlockDevices) < apis.SupportedPRaidType[apis.PoolType(raidGroup.Type)] {
+			return false, fmt.Sprintf("number of block devices honouring raid type should be specified")
+		}
+	}
+	return true, ""
+}
+
+func blockDeviceValidation(bd *apis.CStorPoolClusterBlockDevice) (bool, string) {
+	if bd.BlockDeviceName == "" {
+		return false, fmt.Sprint("block device name cannot be empty")
+	}
+	return true, ""
+}
+
+// validateCSPCDeleteRequest validates CSPC delete request
+// TODO: CSPC delete protection
+func (wh *webhook) validateCSPCDeleteRequest(req *v1beta1.AdmissionRequest) *v1beta1.AdmissionResponse {
+	response := &v1beta1.AdmissionResponse{}
+	response.Allowed = true
+	// TODO: Implement for delete validations
+	return response
+}
+
+// validateCSPCUpdateRequest validates CSPC update request
+// TODO: CSPC delete protection
+func (wh *webhook) validateCSPCUpdateRequest(req *v1beta1.AdmissionRequest) *v1beta1.AdmissionResponse {
```

Comment:
this function seems to be duplicate of validateCSPCCreateRequest.. can we merge them if so?

Reply:
yes